File: grasmere_routes/format.py
"""£ / km / min / % formatters used across the Streamlit pages."""

from __future__ import annotations


def fmt_gbp(pence: int | float | None, digits: int = 2) -> str:
    if pence is None:
        return "—"
    return f"£{pence / 100:,.{digits}f}"


def fmt_gbp_rounded(pence: int | float | None) -> str:
    if pence is None:
        return "—"
    return f"£{pence / 100:,.0f}"


def fmt_km(km: float | None) -> str:
    if km is None:
        return "—"
    return f"{km:,.1f} km"


def fmt_min(m: int | float | None) -> str:
    if m is None:
        return "—"
    m = int(m)
    if m < 60:
        return f"{m}m"
    h, mm = divmod(m, 60)
    return f"{h}h" if mm == 0 else f"{h}h {mm}m"


def fmt_pct(value: float | None, digits: int = 1) -> str:
    if value is None:
        return "—"
    return f"{value * 100:.{digits}f}%"


def fmt_relative_days(days: int | None) -> str:
    if days is None:
        return "never"
    if days == 0:
        return "today"
    if days == 1:
        return "yesterday"
    if days < 30:
        return f"{days} days ago"
    if days < 365:
        return f"{round(days / 30)} months ago"
    return f"{round(days / 365)} years ago"

File: apps/web/lib/format.ts
/** Display helpers for £/km/min/% values. Always render with tabular nums in tables. */
export function fmtGbp(pence: number, opts: { digits?: number } = {}): string {
  const digits = opts.digits ?? 2;
  return new Intl.NumberFormat("en-GB", {
    style: "currency",
    currency: "GBP",
    minimumFractionDigits: digits,
    maximumFractionDigits: digits,
  }).format(pence / 100);
}

export function fmtGbpRounded(pence: number): string {
  return new Intl.NumberFormat("en-GB", {
    style: "currency",
    currency: "GBP",
    maximumFractionDigits: 0,
  }).format(pence / 100);
}

export function fmtKm(km: number): string {
  return `${km.toFixed(1)} km`;
}

export function fmtMin(min: number): string {
  if (min < 60) return `${Math.round(min)}m`;
  const h = Math.floor(min / 60);
  const m = Math.round(min - h * 60);
  return m === 0 ? `${h}h` : `${h}h ${m}m`;
}

export function fmtPct(value: number, digits = 1): string {
  return `${(value * 100).toFixed(digits)}%`;
}

export function fmtRelativeDays(days: number | null): string {
  if (days == null) return "never";
  if (days === 0) return "today";
  if (days === 1) return "yesterday";
  if (days < 30) return `${days} days ago`;
  if (days < 365) return `${Math.round(days / 30)} months ago`;
  return `${Math.round(days / 365)} years ago`;
}
